python_modules/dagster/dagster/_core/pipes/merge_streams.py

```diff
@@ -0,0 +1,232 @@
+        logs: The logs to process.
+        log_handler: The function to call for each log line.
+        shutdown: The threading.Event for signaling that we have to stop processing.
+        interval: The interval on how often to poll the logs if the queue is empty.
+        logger: A simple function to print diagnostic logs.
+
+    """
+    logger.debug(
+        f"Starting to process the merged log stream, will poll the incoming message queue every {interval}s..."
+    )
+    while True:
+        try:
+            # We need to wait for some time for the logs to be put to the priority queue so that they come out as
+            # ordered. If we were to wait on the `logs` queue, we would get the log as soon as it is available and that
+            # is not what we want.
+            entry = logs.get(block=False)
+        except queue.Empty:
+            shutdown.wait(timeout=interval)
+            continue
+
+        try:
+            if entry.log is StopIteration:
+                logger.debug("Shutting down log handler stream...")
+                # This is a special value to finish the processing
+                break
+
+            for line in entry.log.split("\n"):
+                log_handler(line)
+        except Exception as e:
+            logger.warning(f"An error occurred during processing '{entry.log}': {e}")
+        finally:
+            logs.task_done()
+
+
+@contextlib.contextmanager
+def merge_streams(
+    streams: Mapping[str, Generator],
+    log_handler: Callable,
+    stream_processor: Callable[[Iterator[bytes]], Iterator[LogItem]],
+    recent_messages_buffer_size: int = 500,
+    logger: logging.Logger = _default_logger,
+) -> Generator[None, None, None]:
+    """merge_streams and handle them with a single handler.
+
+    The list of streams will be processed in separate threads and the handler will accept
+    each message separately. We expect streams to produce log messages in the format of
+        b'<timestamp> <msg>'
+
+    Args:
+    ----
+        streams: A dict of generators that will provide log messages. It is expected to sometimes observe duplicate
+            messages, which will be deduplicated. The keys to the dictionary are the labels of each stream for
+            debugging purposes.
+        log_handler: The handler to use on each log message returned
+            by the stream.
+        stream_processor: The function that parses the streams into log lines.
+        recent_messages_buffer_size: The size of the buffer for storing the recent messages.
+        logger: The function for logging.
+
+    """
+    workers = queue.Queue()
+    logs = queue.PriorityQueue()
+
+    for label, log_stream in streams.items():
+        workers.put((label, log_stream))
+
+    shutdown = threading.Event()
+    threads = [
+        # Make threads the individual extract logs calls
+        safe_thread(
+            target=_enqueue,
+            kwargs={
+                "streams": workers,
+                "out": logs,
```

Comment:
Note: if you have separate `queue.PriorityQueue` for different outputs (e.g. dagster event stream, stdout, stderr), then here you need to pass all of the priority queues.

The issue is that if there is a race condition, the dagster event stream and stdout/stderr streams could be inconsistent with each other, where the message ordering in different streams will be different, maybe that is a disadvantage of having more than one handling thread.